### PR TITLE
Support for variable length clientDataJSON

### PR DESCRIPTION
**Pull request by akonior**

### CODE REVIEW

File: src/contracts/webauthn_account_contract/src/main.nr [hunk 1]
```diff
@@ -106,14 +106,14 @@ contract WebAuthnAccount {
     global SIGNATURE_LEN = 64;
     global AUTHENTICATOR_DATA_LEN = 37;
     global CHALLENGE_LEN = 43;
-    global CLIENT_DATA_JSON_LEN = 114;
-    global WEBAUTHN_WITNESS_MAX_LEN = 258; // SIGNATURE_LEN + AUTHENTICATOR_DATA_LEN + CHALLENGE_LEN + CLIENT_DATA_JSON_LEN
+    global WEBAUTHN_WITNESS_MAX_LEN = 400; // SIGNATURE_LEN + AUTHENTICATOR_DATA_LEN + CHALLENGE_LEN + CLIENT_DATA_JSON_MAX_LEN + 1
```
Comment (marekkirejczyk) [inline src/contracts/webauthn_account_contract/src/main.nr:109]: Why not explicitly?

File: src/contracts/webauthn_account_contract/src/utils.nr [hunk 2]
```diff
@@ -35,4 +66,12 @@ fn test_convert_array_to_field() {
...
-}
\ No newline at end of file
+}
+
+#[test]
+fn test_f() {
+    let input = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+    assert_eq(std::hash::sha256([1]), prefix_sha256_hash(1, input));
+    assert_eq(std::hash::sha256([1, 2]), prefix_sha256_hash(2, input));
+    assert_eq(std::hash::sha256([1, 2, 3]), prefix_sha256_hash(3, input));
```
Comment (marekkirejczyk) [inline src/contracts/webauthn_account_contract/src/utils.nr:76]: Missing tests for
- [ ] 114
- [ ] 134 
- [ ] 243
- [ ] Length is not supported

File: src/contracts/webauthn_account_contract/src/webauthn.nr [hunk 3]
```diff
@@ -32,49 +36,47 @@ fn concatenate_arrays(a: [u8; 37], b: [u8; 32]) -> [u8; 69] {
...
+        0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0
+        ];
     let signature = [28,230,224,46,41,42,141,231,248,7,247,21,82,94,44,217,93,77,248,54,128,36,40,88,196,12,178,176,42,234,86,195,56,209,248,103,90,44,146,7,232,1,87,180,139,224,93,183,177,224,36,179,150,83,177,224,18,81,127,10,185,237,17,203];
-    validate_webauthn_signature(authenticator_data, client_data_json, signature, challenge, pub_key_x, pub_key_y);
+    validate_webauthn_signature(authenticator_data, 134, client_data_json, signature, challenge, pub_key_x, pub_key_y);
 }
 
 #[test]
-fn test_chrome_challenge32_clientDataJSON_len_243() {
+fn test_chrome_clientDataJSON_len_243() {
```
Comment (marekkirejczyk) [inline src/contracts/webauthn_account_contract/src/webauthn.nr:70]: Missing tests:
- Brave
- Firefox
- Edge

File: src/contracts/webauthn_account_contract/src/main.nr [hunk 4]
```diff
@@ -125,10 +125,11 @@ contract WebAuthnAccount {
             // TODO use challenge from params
```
Comment (marekkirejczyk) [inline src/contracts/webauthn_account_contract/src/main.nr:125]: I wonder what this one is about? Would you elaborate?

File: src/contracts/webauthn_account_contract/src/utils.nr [hunk 5]
```diff
@@ -19,6 +19,37 @@ fn convert_to_array_to_field(a: [u8; 32]) -> Field {
     res
 }
 
+pub fn prefix_sha256_hash<N>(prefix_length: u8, input: [u8; N]) -> [u8; 32] {
```
Comment (marekkirejczyk) [inline src/contracts/webauthn_account_contract/src/utils.nr:22]: It feels off, like it should be more like:
```
    if prefix_length == 2 {
        result = std::hash::sha256(prefix(input, [0; 2])) ;
    }
```

File: src/contracts/webauthn_account_contract/src/main.nr (hunk 1 above)
Comment (akonior) [reply src/contracts/webauthn_account_contract/src/main.nr:109]: Noir limitation
```
error: Unexpected +, expected one of {, ::
    ┌─ /Users/akonior/projects/zk/Quetzal/src/contracts/webauthn_account_contract/src/main.nr:109:53
    │
109 │     global WEBAUTHN_WITNESS_MAX_LEN = SIGNATURE_LEN + AUTHENTICATOR_DATA_LEN + CHALLENGE_LEN + CLIENT_DATA_JSON_MAX_LEN + 1;
    │                                                     -
    │
```

File: src/contracts/webauthn_account_contract/src/utils.nr (hunk 2 above)
Comment (akonior) [reply src/contracts/webauthn_account_contract/src/utils.nr:76]: Tests have been added.

File: src/contracts/webauthn_account_contract/src/utils.nr (hunk 5 above)
Comment (akonior) [reply src/contracts/webauthn_account_contract/src/utils.nr:22]: Repeating `std::hash::sha256` feels less DRY IMO but let's try how it looks. Changed.

Trying to use the latter version results in:
`Message:  Length of array is unknown when evaluating numeric generic`

File: src/contracts/webauthn_account_contract/src/main.nr (hunk 4 above)
Comment (akonior) [reply src/contracts/webauthn_account_contract/src/main.nr:125]: There is a `challenge: Field` param, and in final implementation this should be used and converted to `[u8; 32]`

There is a task for this: https://github.com/orgs/zkworks-xyz/projects/2?pane=issue&itemId=43423555

File: src/contracts/webauthn_account_contract/src/utils.nr (hunk 5 above)
Comment (marekkirejczyk) [reply src/contracts/webauthn_account_contract/src/utils.nr:22]: Looks great!

Comment (marekkirejczyk) [reply src/contracts/webauthn_account_contract/src/utils.nr:22]: Perhaps could be:
```
    if prefix_length == 1 {
        result = prefix(input, [0; 1]);
    }
    if prefix_length == 2 {
        result = prefix(input, [0; 2]);
    }
    if prefix_length == 3 {
        result = prefix(input, [0; 3]);
    }
    if prefix_length == 114 {
        result = prefix(input, [0; 114]);
    }
    if prefix_length == 134 {
        result = prefix(input, [0; 134]);
    }
    if prefix_length == 243 {
        result = prefix(input, [0; 243]);
    }
    std::hash::sha256(result)
```

**Review by marekkirejczyk — APPROVED** [review]

File: src/contracts/webauthn_account_contract/src/webauthn.nr (hunk 3 above)
Comment (akonior) [reply src/contracts/webauthn_account_contract/src/webauthn.nr:72]: Brave is the same as Chrome. Edge not works on Windows (looks like different API), Firefox only supports hardware authenticators.

File: src/contracts/webauthn_account_contract/src/utils.nr (hunk 5 above)
Comment (akonior) [reply src/contracts/webauthn_account_contract/src/utils.nr:22]: Unfortunately this won't work. `prefix(input, [0; 1])` has a different type than `prefix(input, [0; 2])`

File: src/contracts/webauthn_account_contract/src/webauthn.nr (hunk 3 above)
Comment (marekkirejczyk) [reply src/contracts/webauthn_account_contract/src/webauthn.nr:72]: Good to know!

**Review by marekkirejczyk — APPROVED** [review]

File: src/contracts/webauthn_account_contract/src/utils.nr (hunk 5 above)
Comment (marekkirejczyk) [reply src/contracts/webauthn_account_contract/src/utils.nr:22]: I wonder if it could be generic type... Anyhow looks good!

**Pull request merged**